src/librustc_codegen_ssa/back/link.rs

```diff
@@ -968,7 +968,80 @@ pub fn print_native_static_libs(sess: &Session, all_native_libs: &[NativeLibrary
     }
 }
 
+// Because windows-gnu target is meant to be self-contained for pure Rust code it bundles
+// own mingw-w64 libraries. These libraries are often not compatible with mingw-w64
+// installed in the system. This breaks many cases where Rust is mixed with other languages
+// (e.g. *-sys crates).
+// We prefer system mingw-w64 libraries if they are available to avoid this issue.
+fn get_crt_libs_path(sess: &Session) -> Option<PathBuf> {
+    fn find_exe_in_path<P>(exe_name: P) -> Option<PathBuf>
+    where
+        P: AsRef<Path>,
+    {
+        env::var_os("PATH").and_then(|paths| {
+            env::split_paths(&paths)
+                .filter_map(|dir| {
+                    let full_path = dir.join(&exe_name);
+                    if full_path.is_file() {
+                        Some(fix_windows_verbatim_for_gcc(&full_path))
+                    } else {
+                        None
+                    }
+                })
+                .next()
+        })
+    }
+
+    use std::sync::Mutex;
+    lazy_static::lazy_static! {
+        static ref SYSTEM_LIBS: Mutex<Option<Result<PathBuf, ()>>> = Mutex::new(None);
+    }
+
+    let system_libs = SYSTEM_LIBS.lock().unwrap().clone();
+    match system_libs {
+        Some(Ok(compiler_libs_path)) => Some(compiler_libs_path),
+        Some(Err(_)) => None,
+        _ => {
+            if let (linker, LinkerFlavor::Gcc) = linker_and_flavor(&sess) {
+                let linker_path = if cfg!(windows) { linker.with_extension("exe") } else { linker };
```

Comment:
I'm not sure that the `with_extension` here may be right, since this could run the risk of, for example, overriding `*.bat` inputs or something like that? Is there perhaps a `"gcc"` strings elsewhere in the compiler which should instead be `gcc.exe`? Or could this `with_extension` get placed more closely to wherever the string `"gcc"` comes from?

Reply:
It comes from https://github.com/rust-lang/rust/blob/67fae2241ba7d07f455754dcca2c90327b16c098/src/librustc_target/spec/windows_base.rs#L44 which is returned by `linker_and_flavor` without changes. Changing it to `gcc.exe` in the spec would be wrong for cross compilation.

Reply:
That makes sense, yeah, but I think that we should only use `with_extension` in a way that doesn't override user inputs.